### DISCUSSION TITLE
An atomic operation can be optimized?

### DISCUSSION THREAD

**Discussion by wangrunji0408** [tokio-rs/tokio#2940] [Ideas] [answered]
When I perf a program based on Tokio threaded scheduler on ARM64 Linux, I found this atomic operation consumes a lot of time:

https://github.com/tokio-rs/tokio/blob/2e05399f4b6be41680036cc158348973689840ca/tokio/src/runtime/thread_pool/idle.rs#L136

It can be optimized to an atomic load, which prevents cache invalidation on other cores:

```rust
let state = State(self.state.load(SeqCst));
```

After that the overhead of `Idle::worker_to_notify` reduced from 10% to 4% on my workload.

And I'm not sure if it is safe to relax the memory consistency from `SeqCst` to `Acquire` and `Release`. If possible, it will improve the performance of ISAs with weak memory model, such as ARM.

Is that reasonable?

**Comment by sfackler** [comment, marked as answer]
That fetch_add is there intentionally: https://github.com/tokio-rs/tokio/pull/1788
